edx/analytics/tasks/course_blocks.py

```diff
@@ -0,0 +1,501 @@
+            else:
+                no_sort_idx = len(blocks.keys())
+
+            def order_by_sort_idx(key, blocks=blocks, default_sort_idx=no_sort_idx):
+                """Function to sort the blocks on sort_idx"""
+                return blocks[key].get('sort_idx', default_sort_idx)
+
+            for block_id in sorted(blocks.keys(), key=order_by_sort_idx):
+                block = blocks[block_id]
+                is_root = (block_id == root_id)
+                parents = block.get('parents', [])
+                has_multiple_parents = block.get('has_multiple_parents', False)
+                course_path = u''
+                sort_idx = block.get('sort_idx', no_sort_idx)
+
+                if has_multiple_parents:
+                    course_path = unicode(self.multiple_parent_blocks_path)
+                elif not is_root:
+                    if len(parents) == 0:
+                        course_path = unicode(self.deleted_blocks_path)
+                    else:
+                        for parent_id in parents:
+                            parent = blocks[parent_id]
+                            if len(course_path) > 0:
+                                course_path += unicode(self.path_delimiter)
+                            course_path += parent['display_name']
+
+                record = CourseBlockRecord(
+                    course_id=course_id,
+                    block_id=block.get('id'),
+                    block_type=block.get('type'),
+                    display_name=block.get('display_name'),
+                    parent_block_id=block.get('parent_block_id'),
+                    is_root=is_root,
+                    has_multiple_parents=has_multiple_parents,
+                    course_path=course_path,
+                    sort_idx=sort_idx,
+                )
+
+                yield record.to_string_tuple()
+
+    def output(self):
+        return get_target_from_url(self.output_root)
+
+    def complete(self):
+        """
+        The current task is complete if no overwrite was requested,
+        and the output_root/_SUCCESS file is present.
+        """
+        return (super(CourseBlocksApiDataTask, self).complete() and
+                get_target_from_url(url_path_join(self.output_root, '_SUCCESS')).exists())
+
+    def run(self):
+        """
+        Clear out output if data is incomplete, or if overwrite requested.
+        """
+        if not self.complete():
+            self.remove_output_on_overwrite()
+        super(CourseBlocksApiDataTask, self).run()
+
+    def _index_children(self, root_id, blocks, sort_idx=0, more=None):
+        """
+        Applies a sort_idx and parents list to all the blocks in the list.
+        Blocks are sorted in "tree traversal" order: https://en.wikipedia.org/wiki/Tree_traversal
+        """
+        root = blocks.get(root_id, {})
+
+        if more is None:
+            more = []
+
+        children = root.get('children', [])
+        more += children
```

Comment:
This will result in us visiting the same node multiple times if it has multiple parents (and visit all its children again). Possibly add a `visited` set to track which nodes we've been to?